### PR TITLE
Protocol for cache, lru cache and lru cache with ttl (tlru)

### PR DESCRIPTION
**Pull request by jdc123**

### CODE REVIEW

File: src/fixpoint/cache/protocol.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        """Set an item by key"""
+
+    def delete(self, key: K_contra) -> None:
+        """Delete an item by key"""
+
+    def clear(self) -> None:
+        """Clear all items from the cache"""
+
+
+class LRUCache(Protocol[K_contra, V]):
```
Comment (dbmikus) [inline src/fixpoint/cache/protocol.py:26]: If you want to make a protocol for the different cache types, then I think this should be `class LRUCache(Cache[K_contra, V], Protocol[K_contra, V])`.

The LRUCache matches the Cache protocol, and then we want it to be a protocol in itself, so we must also extend `Protocol`.

Then you only need to add class members that are different from the underlying `Cache` class.

File: src/fixpoint/cache/protocol.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+class TLRUCache(Protocol[K_contra, V]):
+    """Protocol for a Time-Limited LRU cache"""
+
+    def get(self, key: K_contra) -> V:
+        """Retrieve an item by key if it has not expired, updating LRU order"""
+
+    def set(self, key: K_contra, value: V, ttl: int) -> None:
+        """Set an item by key with a time-to-live (TTL)"""
+
+    def expire(self) -> None:
```
Comment (dbmikus) [inline src/fixpoint/cache/protocol.py:51]: I feel like this should be an internal method. Why do we want our caller to be responsible for expiring cache members?

**Comment by dbmikus** [comment]
Oh also, just curious, why accept contravariant vs covariant types? aka super-types vs subtypes? I don't really have much of an opinion on which one to accept, but was interested if you had any insight (or if we just found that somewhere and decided to follow it!)

File: src/fixpoint/cache/protocol.py (hunk 2 above)
Comment (jdc123) [reply src/fixpoint/cache/protocol.py:51]: Fair point

**Comment by jdc123** [comment]
@dbmikus Regarding contravariant vs covariant, I was getting typecheck errors when I wasn't specifying contravariant vs covariant. I went off of a chat bot suggestion which yielded:

```The lint error about an invariant type variable "K" used in a protocol where a contravariant one is expected typically arises when a type variable is used in a position that requires contravariance, but the type variable is defined as invariant. In Python's type hinting system, type variables can be invariant, covariant, or contravariant:
Invariant means the type must be exactly as specified (default behavior).
Covariant (covariant=True) means the type can be a specified type or a more derived type.
Contravariant (contravariant=True) means the type can be a specified type or a more general type.
In the context of your cache protocols, the type variable K used for keys should be contravariant because a function that can accept a broader type (a supertype) for keys is more flexible and should be substitutable where a narrower type (a subtype) is expected. This is particularly relevant for methods like get, set, delete, which accept a key as an argument.```

File: src/fixpoint/cache/protocol.py (hunk 1 above)
Comment (jdc123) [reply src/fixpoint/cache/protocol.py:26]: Yeah make sense

File: src/fixpoint/cache/protocol.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        """Retrieve an item by key if it has not expired, updating LRU order"""
+
+    def set(self, key: K_contra, value: V, ttl: int) -> None:
+        """Set an item by key with a time-to-live (TTL)"""
+
+    def delete(self, key: K_contra) -> None:
+        """Delete an item by key"""
+
+    def clear(self) -> None:
+        """Clear all items from the cache"""
```
[the remark this answers is not in the document]
Comment (dbmikus) [reply src/fixpoint/cache/protocol.py:55]: nvm didn't realize the additional TTL argument. `set` is not redundant.

**Pull request merged**